### PR TITLE
BVH improvements

### PR DESCRIPTION
**Pull request by adayton1**
# Summary

- This PR is a feature
- It does the following:
  - Adds a constructor to BVH at the request of myself (see below)
  - Avoids unnecessary copy of each bounding box

## Current pattern
```
axom::spin::BVH<3> bvh;
bvh.setAllocatorID(allocatorID);
bvh.initialize(boxes, numBoxes);
```

## New pattern
```
axom::spin::BVH<3> bvh(boxes, numBoxes, allocatorID);
```

### CODE REVIEW

**Comment by adayton1** [comment]
My only concern is that initialize returns a bool indicating whether construction succeeds. But it looks like it always return true.

**Comment by white238** [comment]
"Avoids unnecessary copy of each bounding box" @adayton1 have you ever thought that maybe we want to extra work?!?! Why do you have to make everything so efficient and better?

File: src/axom/spin/BVH.hpp [hunk 1]
```diff
@@ -194,6 +194,42 @@ class BVH
...
+   * \param [in] allocatorID the ID of the allocator to use in BVH construction.
+   * \param [in] scaleFactor the scale factor for the given bounding boxes.
+   * \param [in] tolerance the tolerance to use when querying the BVH.
+   *
+   * \warning The supplied boxes array must point to a buffer in a memory space
+   *  that is compatible with the execution space. For example, when using
+   *  CUDA_EXEC, boxes must be in unified memory or GPU memory. The supplied
+   *  allocator ID must also correspond to an umpire allocator and be compatible
+   *  with the execution space. The code currently does not check for these
+   *  conditions.
```
Comment (agcapps) [inline src/axom/spin/BVH.hpp:213]: big ol' fat warning

Comment (agcapps) [reply src/axom/spin/BVH.hpp:213]: thank you for that

**Review by agcapps — APPROVED** [review]

File: src/axom/spin/BVH.hpp [hunk 2]
```diff
@@ -194,6 +194,42 @@ class BVH
...
+   *      with the execution space
+   */
+  template <typename BoxIndexable>
+  BVH(const BoxIndexable boxes,
+      IndexType numItems,
+      int allocatorID = axom::execution_space<ExecSpace>::allocatorID(),
+      FloatType tolerance = DEFAULT_TOLERANCE,
+      FloatType scaleFactor = DEFAULT_SCALE_FACTOR)
+    : m_AllocatorID{allocatorID}
+    , m_tolerance{tolerance},
```
Comment (kennyweiss) [inline src/axom/spin/BVH.hpp:227]: Looks like there's an extra comma here. Please remove the one at the end of the line.

File: src/axom/spin/BVH.hpp [hunk 3]
```diff
@@ -194,6 +194,42 @@ class BVH
    */
   BVH() : m_AllocatorID(axom::execution_space<ExecSpace>::allocatorID()) { }
 
+  /*!
+   * \brief Constructs a BVH instance, of specified dimension, over a given set
```
Comment (rhornung67) [inline src/axom/spin/BVH.hpp:198]:
```suggestion
   * \brief Constructs a BVH instance of specified dimension over a given set
```

**Review by rhornung67 — COMMENTED** [review]
One minor comment nit and CI check compilation issues to be fixed.

File: src/axom/spin/BVH.hpp [hunk 4]
```diff
@@ -194,6 +194,42 @@ class BVH
...
+      IndexType numItems,
+      int allocatorID = axom::execution_space<ExecSpace>::allocatorID(),
+      FloatType tolerance = DEFAULT_TOLERANCE,
+      FloatType scaleFactor = DEFAULT_SCALE_FACTOR)
+    : m_AllocatorID{allocatorID}
+    , m_tolerance{tolerance},
+    , m_scaleFactor{scaleFactor}
+  {
+    initialize(boxes, numItems);
+  }
```
Comment (kennyweiss) [inline src/axom/spin/BVH.hpp:231]: If possible, please add a unit test w/ this constructor

**Review by kennyweiss — APPROVED** [review]
Thanks @adayton1 !

File: src/axom/spin/BVH.hpp (hunk 2 above)
Comment (kennyweiss) [reply src/axom/spin/BVH.hpp:227]: CI appears to have found this too!  
https://dev.azure.com/healy22/Status%20Check%20Testiing/_build/results?buildId=10773&view=logs&j=9a6faa8f-a429-59ac-ab99-2d21e7e60d32&t=f778077e-74b2-5f79-ae4c-24dad9f21ef8

**Review by rhornung67 — APPROVED** [review]

**Review by bmhan12 — APPROVED** [review]

**Pull request merged**